tapdb/assets_store.go

```diff
@@ -1425,10 +1469,14 @@ func queryChainAssets(ctx context.Context, q ActiveAssetsStore,
 }
 
 // FetchCommitment returns a specific commitment identified by the given asset
-// parameters. If no commitment is found, ErrNoCommitment is returned.
+// parameters. If no commitment is found, ErrNoCommitment is returned. With
+// mustBeLeased the caller decides whether the asset output should've been
+// leased before or not. If mustBeLeased is false, then the state of the lease
+// is not checked.
 func (a *AssetStore) FetchCommitment(ctx context.Context, id asset.ID,
 	anchorPoint wire.OutPoint, groupKey *asset.GroupKey,
-	scriptKey *asset.ScriptKey) (*tapfreighter.AnchoredCommitment, error) {
+	scriptKey *asset.ScriptKey,
+	mustBeLeased bool) (*tapfreighter.AnchoredCommitment, error) {
```

Comment:
So this is a sort of sanity check we want to implement in the constraints of a DB transaction? What about a similar coarse coin selection mutex to what we have in `lnd`? 

Looking at it from another angle: wouldn't one already fail a coin selection attempt if the outputs needed are already leased? Since first we'd try to perform coin selection with either `ListEligibleCoins` or `SelectForAmount`?

Reply:
The `FetchCommitment` is needed when we go to anchor a signed vPSBT. So at that point the caller MUST have leased the outpoint already.
Flow is:
 1. Fund vPSBT -> Calls `ListEligibleCoins`, then `SelectForAmount`, then adds lease for the "selected" coin. PSBT is returned to caller (or freighter state machine is advanced, but either way, this step is done)
 2. Sign vPSBT -> No DB interaction required
 3. AnchorVirtualPsbts -> Calls `FetchCommitment` to load the inputs required to create the new tree. At this point we select the coin being spent and we need to make sure it was previously leased or is currently still leased.

The other place where `FetchCommitment` is called is for creating an asset ownership proof. We also need to select the full commitment for that, but there we don't care if the asset is currently leased or not.